src/main/scala/com/scality/clueso/query/cache/SessionCacheManager.scala

```diff
@@ -0,0 +1,97 @@
+package com.scality.clueso.query.cache
+
+import java.util.Date
+import java.util.concurrent.atomic.AtomicReference
+
+import com.scality.clueso.CluesoConfig
+import com.scality.clueso.query.MetadataQueryExecutor.setupDf
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.joda.time.DateTime
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.{Await, Future}
+
+object SessionCacheManager extends LazyLogging {
+  var bucketDfs = Map[String, AtomicReference[DataFrame]]()
+  var bucketUpdateTs = Map[String, DateTime]()
+
+  val setupDfLock = new scala.collection.parallel.mutable.ParHashSet[String]()
+
+  def getCachedBucketDataframe(spark: SparkSession, bucketName: String)(implicit config: CluesoConfig): DataFrame = {
+    // check if cache doesn't exist
+    val tableView = s"${new Date().getTime}_$bucketName"
+
+    if (!bucketDfs.contains(bucketName)) {
+      val bucketDf = setupDf(spark, config, bucketName)
+
+      if (acquireLock(bucketName)) {
+        bucketDf.createOrReplaceTempView(tableView)
+        bucketDf.sparkSession.catalog.cacheTable(tableView)
+
+        bucketDfs = bucketDfs.updated(bucketName, new AtomicReference(bucketDf))
+        bucketUpdateTs = bucketUpdateTs.updated(bucketName, DateTime.now())
+        releaseLock(bucketName)
+      }
+
+      bucketDf
+    } else {
+      // cache exists
+      if (bucketUpdateTs(bucketName).plus(config.cacheExpiry.toMillis).isBeforeNow) {
+        // check if there's a dataframe update going on
+        if (acquireLock(bucketName)) {
+
+          // async update dataframe if there's none already doing it
+          Future {
+            logger.info(s"Calculating view $tableView")
+            val bucketDf = setupDf(spark, config, bucketName)
+
+            bucketDf.createOrReplaceTempView(tableView)
+
+            // this operation triggers execution
+            val cachingAsync = Future {
+              bucketDf.sparkSession.catalog.cacheTable(tableView)
+            }
+            import scala.concurrent.duration._
+            Await.ready(cachingAsync, 3 minutes)
+
+            bucketDf
+          } map { bucketDf =>
+            logger.info(s"Calculating view $tableView")
+            bucketDf.count() // force calculation
+            logger.info(s"Atomically swapping RDD for bucket = $bucketName ( new = $tableView )")
+            val oldDf = bucketDfs.getOrElse(bucketName, new AtomicReference()).getAndSet(bucketDf)
+            bucketUpdateTs = bucketUpdateTs.updated(bucketName, DateTime.now())
```

Comment:
does this just keep growing or do you ever remove the added timestamps?

Reply:
The new timestamps for the same bucket override the last ones. However, we have no mechanism of actively expiring caches, no LRU either, let's discuss later